File: hamic_algo/imple_medium/maxProduct.py
def maxProduct(a):
    m = int(1e9+7)
    mx = int(-1e18)
    tich1 = 1
    tich2 = 1
    am1 = 0
    am2 = 0

    for i in range(len(a)):
        tich2 = ((tich2 + m) % m * (a[i] + m) % m) % m
        if (a[i] < 0):
            am2 += 1
        if am2 % 2 == 1:
            tich2 -= m 
        # print(tich2)
    
    for i in range(len(a)-1):
        # if tich1 * a[i] < 0:
        #     tich1 = ((tich1 + m) % m * (a[i] + m) % m) % m
        #     tich1 -= m
        # else:
        if a[i] < 0:
            am1 += 1
        tich1 = ((tich1 + m) % m * (a[i] + m) % m) % m
        if (am1 % 2 == 1):
            tich1 -= m

        # if tich2 * a[i] < 0:
        #     tich2 = (tich2 + m) % m * pow(a[i] + m, m-2, m) % m
        #     tich2 -= m
        # else:
        if a[i] < 0:
            am2 -= 1
        tich2 = (tich2 + m) % m * pow(a[i] + m, m-2, m) % m
        if am2 % 2 == 1:
            tich2 -= m

        tmp1 = min(tich1, tich2)
        tmp2 = max(tich1, tich2)
        if (tmp1 * tmp2 > 0 and tmp1 < 0) or tmp1 < -tmp2:
            mx = max(mx, (tmp1 + tmp2) % m - m)
        else:
            mx = max(mx, (tich1 + tich2) % m)
        # if (tich1 * tich2 > 0 and tich1 < 0):
        #     mx = max(mx, (tich1 + tich2) % m - m)
        # else:
        #     mx = max(mx, (tich1 + tich2) % m)
        
        # print(tich1, tich2)
        # print(mx)
    return mx

a = [-3,-2,-2,7,3]
# for i in range(0, int(1e4)):
#     a.append(int(1e4))
print(maxProduct(a))

File: hamic_algo/imple_medium/maxProduct.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long

ll m = 1e9+7;
ll maxProduct(vector<ll> a) {
    ll mx = -1e9;
    ll tich1 = 1;
    ll tich2 = 1;
    int n = a.size();
    vector<ll> v1(n, 1), v2(n, 1);
    v1[0] = a[0];
    for (int i = 1; i < n; ++i) {
        v1[i] = v1[i-1] * a[i] % m;
    }
    v2[n-1] = a[n-1];
    for (int i = n-2; i > -1; --i) {
        v2[i] = v2[i+1] * a[i] % m;
    }
    for(int i = 1; i < n; ++i) {
        mx = max(mx, (v1[i-1] + v2[i]) % m);
    } 
    return mx;
}

int main(){
    vector<ll> a = {-1,3,4,-2}; 
    cout << maxProduct(a);
    // cout << power(-1, 4);
}